File: starters/java/src/main/java/io/fproto/core/InvalidTransitionException.java
package io.fproto.core;

public class InvalidTransitionException extends Exception {
    private final SessionStatus status;
    private final SessionEvent event;

    public InvalidTransitionException(SessionStatus status, SessionEvent event) {
        super("invalid transition: " + status + " + " + event);
        this.status = status;
        this.event = event;
    }

    public SessionStatus getStatus() {
        return status;
    }

    public SessionEvent getEvent() {
        return event;
    }
}

File: starters/php/src/Session/InvalidTransitionException.php
<?php

declare(strict_types=1);

namespace FProto\Session;

use RuntimeException;

final class InvalidTransitionException extends RuntimeException
{
    public function __construct(
        public readonly SessionStatus $status,
        public readonly SessionEvent $event,
    ) {
        parent::__construct(
            sprintf('invalid transition: %s + %s', $status->value, $event->value)
        );
    }
}
